sdk/storage/azdatalake/sas/service.go

```diff
@@ -0,0 +1,407 @@
+	p.signedTID = *udk.SignedTID
+	p.signedStart = *udk.SignedStart
+	p.signedExpiry = *udk.SignedExpiry
+	p.signedService = *udk.SignedService
+	p.signedVersion = *udk.SignedVersion
+
+	return p, nil
+}
+
+// getCanonicalName computes the canonical name for a container or blob resource for SAS signing.
+func getCanonicalName(account string, filesystemName string, fileName string, directoryName string) string {
+	// Container: "/blob/account/containername"
+	// Blob:      "/blob/account/containername/blobname"
+	elements := []string{"/blob/", account, "/", filesystemName}
+	if fileName != "" {
+		elements = append(elements, "/", strings.Replace(fileName, "\\", "/", -1))
+	} else if directoryName != "" {
+		elements = append(elements, "/", directoryName)
+	}
+	return strings.Join(elements, "")
+}
+
+// FilesystemPermissions type simplifies creating the permissions string for an Azure Storage container SAS.
+// Initialize an instance of this type and then call its String method to set BlobSignatureValues' Permissions field.
+// All permissions descriptions can be found here: https://docs.microsoft.com/en-us/rest/api/storageservices/create-service-sas#permissions-for-a-directory-container-or-blob
+type FilesystemPermissions struct {
+	Read, Add, Create, Write, Delete, List, Move bool
+	Execute, ModifyOwnership, ModifyPermissions  bool // Meant for hierarchical namespace accounts
+}
+
+// String produces the SAS permissions string for an Azure Storage container.
+// Call this method to set BlobSignatureValues' Permissions field.
+func (p *FilesystemPermissions) String() string {
+	var b bytes.Buffer
+	if p.Read {
+		b.WriteRune('r')
+	}
+	if p.Add {
+		b.WriteRune('a')
+	}
+	if p.Create {
+		b.WriteRune('c')
+	}
+	if p.Write {
+		b.WriteRune('w')
+	}
+	if p.Delete {
+		b.WriteRune('d')
+	}
+	if p.List {
+		b.WriteRune('l')
+	}
+	if p.Move {
+		b.WriteRune('m')
+	}
+	if p.Execute {
+		b.WriteRune('e')
+	}
+	if p.ModifyOwnership {
+		b.WriteRune('o')
+	}
+	if p.ModifyPermissions {
+		b.WriteRune('p')
+	}
+	return b.String()
+}
+
+// Parse initializes ContainerPermissions' fields from a string.
+func parseFilesystemPermissions(s string) (FilesystemPermissions, error) {
+	p := FilesystemPermissions{} // Clear the flags
+	for _, r := range s {
+		switch r {
+		case 'r':
+			p.Read = true
+		case 'a':
+			p.Add = true
+		case 'c':
+			p.Create = true
+		case 'w':
+			p.Write = true
+		case 'd':
+			p.Delete = true
+		case 'l':
+			p.List = true
+		case 'm':
+			p.Move = true
+		case 'e':
+			p.Execute = true
+		case 'o':
+			p.ModifyOwnership = true
+		case 'p':
+			p.ModifyPermissions = true
+		default:
+			return FilesystemPermissions{}, fmt.Errorf("invalid permission: '%v'", r)
+		}
+	}
+	return p, nil
+}
+
+// FilePermissions type simplifies creating the permissions string for an Azure Storage blob SAS.
+// Initialize an instance of this type and then call its String method to set BlobSignatureValues' Permissions field.
+type FilePermissions struct {
```

Comment:
Should this be Path Permissions?

Reply:
I dont see a DirectoryPermissions here, am I missing something?

Reply:
whoa, idk how I deleted it before pushing... let me add it again haha

Reply:
but yeah anyway, Java uses the PathPermissions type and .NET calls it DatalakePermissions, so ultimately it is a design decision here

Reply:
I'll go for PathPermissions actually.